### PR TITLE
Revise and update mappings

### PR DESCRIPTION
**Pull request by nickadamson**
This pull request updates the Subgraph's mappings & handlers to use the new schema introduced in #44.

I have documented most of the existing functions to the best of my understanding and the new ones entirely.

[The biggest change is how the subgraph handles metrics.](https://github.com/valorem-labs-inc/valorem-subgraph/blob/91f43f53cc358e347e62adf02704241cbe894ea1/src/utils/index.ts#L114) The logic of this function should be reviewed by a few of us in a meeting to ensure there are no errors.

[There is a nightly deployment of this subgraph live and emitting logs with how the metrics change on a transactional basis on the Graph.](https://thegraph.com/hosted-service/subgraph/nickadamson/test-goerli2?selected=logs)

[There is also a preview deployment of the frontend using this nightly subgraph's as the data source.](https://valorem-frontend-f1j9vqtu0-valorem.vercel.app/metrics)

There are a few testnet related lines that should be deleted for mainnet release. They are being tracked in #34

### CODE REVIEW

File: src/utils/price.ts [hunk 1]
```diff
@@ -42,7 +45,10 @@ export function getEthPriceInUSD(): BigDecimal {
   );
 
   if (
-    daiPool.token0().toHexString().toLowerCase() == WETH_ADDRESS.toLowerCase()
+    daiPool
```
Comment (0xAlcibiades) [inline src/utils/price.ts:48]: Could this stuff be cased on the chain ID? Seems like we need it.

File: src/utils/index.ts [hunk 2]
```diff
@@ -1,57 +1,424 @@
...
+    exerciseDaily.notionalVolSum = exerciseDaily.notionalVolSum.plus(
+      exerciseAmountTotal
+    );
+    exerciseDaily.notionalVolSumUSD = exerciseDaily.notionalVolSumUSD.plus(
+      exerciseTotalUSD
+    );
   }
+  exerciseDaily.save();
+  underlyingDaily.save();
+  // TODO: Remove logs before merging
```
Comment (0xAlcibiades) [inline src/utils/index.ts:374]: Remove these or turn down to debug

**Review by 0xAlcibiades — CHANGES_REQUESTED** [review]

File: src/utils/price.ts (hunk 1 above)
Comment (nickadamson) [reply src/utils/price.ts:48]: created #47 to track this. will include as part of upgrading the repo to support multiple chains

File: src/utils/index.ts (hunk 2 above)
Comment (nickadamson) [reply src/utils/index.ts]: removed :+1:

**Review by 0xAlcibiades — APPROVED** [review]

File: src/constants.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+// Credit to https://github.com/OpenZeppelin/openzeppelin-subgraphs
```
Comment (neodaoist) [inline src/constants.ts:1]: Should call out the license too, MIT in this case

File: src/fetch/account.ts [hunk 4]
```diff
@@ -1,10 +1,16 @@
...
-import { Address } from "@graphprotocol/graph-ts";
-
 import { Account } from "../../generated/schema";
 
-export function fetchAccount(address: Address): Account {
-  let account = new Account(address.toHex());
+/**
+ * Searches for and returns an Account, initializing a new one if not found
+ * @param {string} address.toHexString()
+ * @return {*}  {Account}
```
Comment (neodaoist) [inline src/fetch/account.ts:7]: Is there an extra space between `}` and `{` ? Not sure on formatting conventions but wanted to just check, here and elsewhere with the similar notation.

File: src/mapping.ts [hunk 5]
```diff
@@ -1,410 +1,477 @@
...
-  loadOrInitializeFeeSwitch,
-  loadOrInitializeToken,
-  updateTokenDayData,
-  updateValoremDayData,
+  fetchOptionSettlementEngine,
+  fetchToken,
+  fetchDailyTokenMetrics,
+  fetchDailyOSEMetrics,
+  handleDailyMetrics,
+  RedeemOrTransferAmounts,
```
Comment (neodaoist) [inline src/mapping.ts:42]: Should "Redeem" be lower-cased here, to match the others?

File: src/mapping.ts [hunk 6]
```diff
@@ -1,410 +1,477 @@
...
-  exerciseAsset.save();
-
-  let dayData = updateValoremDayData(event);
-
-  dayData.volumeUSD = dayData.volumeUSD.plus(underlyingValueUSD);
+  dailyOSEMetrics.save();
+}
 
-  dayData.save();
+// https://github.com/OpenZeppelin/openzeppelin-subgraphs
```
Comment (neodaoist) [inline src/mapping.ts:304]: Standardize here and elsewhere how we include and credit code+license type. This doesn't match the first few.

File: src/mapping.ts [hunk 7]
```diff
@@ -1,410 +1,477 @@
...
-  const underlyingPriceUSD = getTokenPriceUSD(underlyingAssetAddress as string);
+// https://github.com/OpenZeppelin/openzeppelin-subgraphs
+export function handleTransferBatch(event: TransferBatchEvent): void {
+  const contract = fetchERC1155(event.address.toHexString());
+  const operator = fetchAccount(event.params.operator.toHexString());
+  const from = fetchAccount(event.params.from.toHexString());
+  const to = fetchAccount(event.params.to.toHexString());
+  const ids = event.params.ids;
+  const amounts = event.params.amounts;
+  // If this equality doesn't hold (some devs actually don't follow the ERC specifications) then we just can't make
```
Comment (neodaoist) [inline src/mapping.ts:357]: Not sure we need to keep this salty OZ comment in =)

File: src/mapping.ts [hunk 8]
```diff
@@ -1,410 +1,477 @@
...
+    operator,
+    from,
+    to,
+    event.params.id,
+    event.params.amount
   );
-  underlyingDayData.save();
 
-  updateTokenDayData(exerciseAsset, event);
+  /**
```
Comment (neodaoist) [inline src/mapping.ts:322]: I really like how you setup this pattern FWIW with the 0x0 checks and the dedicated function for handling ERC1155 transfer metrics.

File: src/mapping.ts [hunk 9]
```diff
@@ -1,410 +1,477 @@
...
+  const fromAddress = event.params.from.toHexString();
+  const toAddress = event.params.to.toHexString();
+  const tokenId = event.params.id;
+  const amount = event.params.amount;
+
+  // check transfer isn't an OSE Write/Exercise/Redeem event (mint/OSE is receiver/burn)
+  const transferIsExternal =
+    fromAddress != ZERO_ADDRESS &&
+    toAddress != ZERO_ADDRESS &&
+    toAddress != event.address.toHexString();
```
Comment (neodaoist) [inline src/mapping.ts:337]: Hmm, have to think this through, but what is the scenario where an 1155 is getting transferred _to_ the OSE contract itself?

File: src/mapping.ts [hunk 10]
```diff
@@ -1,410 +1,477 @@
...
-    .div(
-      exponentToBigDecimal(
-        BigInt.fromI64(
-          ERC20.bind(
-            Address.fromString(underlyingAssetAddress as string)
-          ).decimals()
-        )
-      )
-    );
+  /**
```
Comment (neodaoist) [inline src/mapping.ts:375]: Anyway we can de-dupe this Valorem Extension code to DRY things up?

File: src/utils/index.ts [hunk 11]
```diff
@@ -1,57 +1,372 @@
...
+   *  Update Daily OSE Metrics
+   */
+  // By saving the before/after TVL in USD, we are able to maintain consistent records
+  dailyOSEMetrics.totalValueLockedUSD = dailyOSEMetrics.totalValueLockedUSD
+    .minus(underlyingTVLUSDBefore)
+    .minus(exerciseTVLUSDBefore)
+    .plus(underlyingTVLUSDAfter)
+    .plus(exerciseTVLUSDAfter);
+
+  if (eventKind == "write") {
```
Comment (neodaoist) [inline src/utils/index.ts:194]: Not entirely sure how this flow control works, but could it be more efficient to use if/elseif/else block?

File: src/utils/price.ts [hunk 12]
```diff
@@ -23,6 +23,9 @@ const TOKEN_WHITELIST = [
 
 // Gets ETHs price in USD using the DAI / WETH Uniswap V3 pool.
 export function getEthPriceInUSD(): BigDecimal {
+  // TODO: REVERT FOR MAINNET. GOERLI POOLS ARE INACCURATE
+  // ! hardcode ETH price 1250 USD
+  return BigDecimal.fromString("1250");
```
Comment (neodaoist) [inline src/utils/price.ts:28]: Maybe we could update to 1500 if going to be testing soon =)

**Review by neodaoist — CHANGES_REQUESTED** [review]

File: src/constants.ts (hunk 3 above)
Comment (nickadamson) [reply src/constants.ts:1]: Will add

File: src/fetch/account.ts (hunk 4 above)
Comment (nickadamson) [reply src/fetch/account.ts:7]: Hm, not 100% on that, its a VSCode prompt called Document This, and that is how it generated it.

File: src/mapping.ts (hunk 5 above)
Comment (nickadamson) [reply src/mapping.ts:42]: RedeemOrTransfer amounts is a type/struct, so in this case it should be capitalized.

File: src/mapping.ts (hunk 6 above)
Comment (nickadamson) [reply src/mapping.ts:304]: 👍

File: src/mapping.ts (hunk 7 above)
Comment (nickadamson) [reply src/mapping.ts:357]: HA I agree.

File: src/mapping.ts (hunk 8 above)
Comment (nickadamson) [reply src/mapping.ts:322]: <3 Thank you

File: src/mapping.ts (hunk 9 above)
Comment (nickadamson) [reply src/mapping.ts:337]: Honestly, there shouldn't be. And it's not like the OSE can transfer them out either. Might remove this check.

File: src/mapping.ts (hunk 10 above)
Comment (nickadamson) [reply src/mapping.ts:375]: I tried to keep the notice close to the changes, so in the future anyone looking will be able to tell what's OZ/what's Valorem's. What do you have in mind?

File: src/utils/index.ts (hunk 11 above)
Comment (nickadamson) [reply src/utils/index.ts:194]: Ideally, I would use a switch statement, but that's not a thing in current state of AssemblyScript. I'll refactor to use if/else's though.

File: src/utils/price.ts (hunk 12 above)
Comment (nickadamson) [reply src/utils/price.ts:28]: 🧠 🤝

File: src/constants.ts (hunk 3 above)
Comment (neodaoist) [reply src/constants.ts:1]: // Credit to <URL>, included under <LICENSE>

File: src/mapping.ts (hunk 10 above)
Comment (neodaoist) [reply src/mapping.ts:375]: - DRY up the code itself.
- Rename to `isMintOrBurn(from, to)`.

File: src/constants.ts (hunk 3 above)
Comment (nickadamson) [reply src/constants.ts:1]: done in https://github.com/valorem-labs-inc/valorem-subgraph/commit/ecb0958c55a60effddb94cc5d773b72580758013

File: src/mapping.ts (hunk 6 above)
Comment (nickadamson) [reply src/mapping.ts:304]: done in https://github.com/valorem-labs-inc/valorem-subgraph/commit/ecb0958c55a60effddb94cc5d773b72580758013

File: src/fetch/account.ts (hunk 4 above)
Comment (nickadamson) [reply src/fetch/account.ts:7]: done in https://github.com/valorem-labs-inc/valorem-subgraph/commit/c30b7972063c3b6bec983284f757d9febe3f7a67

File: src/mapping.ts (hunk 7 above)
Comment (nickadamson) [reply src/mapping.ts:357]: removed https://github.com/valorem-labs-inc/valorem-subgraph/commit/c09a1c3b1e5fee3131700cb87bbe0fe0d9804827

File: src/mapping.ts (hunk 9 above)
Comment (nickadamson) [reply src/mapping.ts:337]: removed OSE check here: https://github.com/valorem-labs-inc/valorem-subgraph/commit/59eefb0dc693d6b4f3a5d88582dcd7335a26a7a4

File: src/mapping.ts (hunk 10 above)
Comment (nickadamson) [reply src/mapping.ts:375]: refactored here: https://github.com/valorem-labs-inc/valorem-subgraph/commit/59eefb0dc693d6b4f3a5d88582dcd7335a26a7a4

File: src/utils/index.ts (hunk 11 above)
Comment (nickadamson) [reply src/utils/index.ts:194]: elifs added https://github.com/valorem-labs-inc/valorem-subgraph/commit/0c078e644a7c38e9ff7af689a1784ff579a6112f

File: src/utils/price.ts (hunk 12 above)
Comment (nickadamson) [reply src/utils/price.ts:28]: 1550 ;) https://github.com/valorem-labs-inc/valorem-subgraph/commit/7480f4849e93de9ff47666b010f1c62a5d88499d

**Pull request merged**